lib/filterx/object-datetime.c

```diff
@@ -255,10 +255,10 @@ _strptime_eval(FilterXExpr *s)
   const gchar *time_str;
   gsize time_str_len;
   gboolean extract_success = filterx_object_extract_string(time_str_obj, &time_str, &time_str_len);
-  filterx_object_unref(time_str_obj);
 
   if (!extract_success)
     {
+      filterx_object_unref(time_str_obj);
```

Comment:
We have a lot of `filterx_object_extract_*()` functions, the naming for most of them seems okay, except for `string`, `protobuf`, and `bytes`. For those types, the extracted value is not really extracted but borrowed.
That may have been one of the reasons why this use-after-free was introduced.

I would rename these functions to denote this difference. If we insist on keeping the `extractor` idea in the name, adding `ref` may work too, for example, `filterx_object_extract_string_ref()`.

Reply:
Just chiming in, how about `filterx_object_as_string()`?
(Inspired by [Rust conventions](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv))

Reply:
Good idea. Added `ref` to several getter functions that have this behavior.

Reply:
To me `filterx_object_as_string()` still does not communicate clearly that it is a borrowed reference to the underlying string. Somehow we should incorporate `ref` to the function name.

Reply:
Oh my bad the "as" word communicates it.